optuna/samplers/nsgaiii.py

```diff
@@ -0,0 +1,554 @@
+
+    # count_reference_points keeps how many elite neighbors each reference point has
+    count_reference_points: Dict[int, int] = DefaultDict(int)
+    for i, reference_point_id in enumerate(closest_reference_points[:elite_population_num]):
+        count_reference_points[reference_point_id] += 1
+
+    # reference_point_to_population keeps pairs of a neighbor and the distance of each reference
+    # point from borderline front population
+    reference_point_to_population = DefaultDict(list)
+    for i, reference_point_id in enumerate(closest_reference_points[elite_population_num:]):
+        reference_point_to_population[reference_point_id].append(
+            (distance_reference_points[i + elite_population_num], i)
+        )
+
+    # nearest_points_count_to_reference_point classifies reference points which have at least one
+    # closest borderline population member by the number of elite neighbors they have. Each key
+    # correspond to the count of elite neighbors and values are reference point ids.
+    nearest_points_count_to_reference_point = DefaultDict(list)
+    for reference_point_id in reference_point_to_population:
+        count = count_reference_points[reference_point_id]
+        nearest_points_count_to_reference_point[count].append(reference_point_id)
+
+    return nearest_points_count_to_reference_point, reference_point_to_population
+
+
+def _preserve_niche_individuals(
+    target_population_size: int,
+    population: List[FrozenTrial],
+    nearest_points_count_to_reference_point: Dict[int, List[int]],
+    reference_point_to_population: Dict[int, List[Tuple[float, int]]],
+    rng: np.random.RandomState,
+) -> List[FrozenTrial]:
+    """Determine who survives form the borderline front
+
+    Who survive form the borderline front is determined according to the sparsity of each closest
+    reference point. The algorithm picks a reference point from those who have the least neighbors
+    in elite population and adds one of borderline front member who has the same closest reference
+    point.
+    """
+
+    count = 0
+    additional_elite_population: List[FrozenTrial] = []
+    while len(additional_elite_population) < target_population_size:
+        if not nearest_points_count_to_reference_point[count]:
+            count += 1
+            continue
+
+        individual_idx = rng.choice(len(nearest_points_count_to_reference_point[count]))
+        (
+            nearest_points_count_to_reference_point[count][individual_idx],
+            nearest_points_count_to_reference_point[count][-1],
+        ) = (
+            nearest_points_count_to_reference_point[count][-1],
+            nearest_points_count_to_reference_point[count][individual_idx],
+        )
+
+        reference_point_id = nearest_points_count_to_reference_point[count].pop()
+        if count:
+            reference_point_idx = rng.choice(
+                len(reference_point_to_population[reference_point_id])
```

Comment:
What happen if `reference_point_to_population[reference_point_id]` is an empty set?

Reply:
`reference_point_to_population[reference_point_id]` won't be empty because
- `nearest_points_count_to_reference_point` only has reference points with non-empty neighbors list at first.
-  `reference_point_id` is pushed back to `nearest_point_count_to_reference_point` again after one of the neighbor is selected as a member elite population and deleted from the neighbors list only when the list is non-empty.